### ISSUE TITLE
Gnuk cert import: Transmit failed

### ISSUE THREAD

**Issue by alex-nitrokey** [OpenSC/OpenSC#1282]
### Problem Description
While trying to follow the [wiki](https://github.com/OpenSC/OpenSC/wiki/OpenPGP-card#6-import-key-resp-certificate) instructions to import a certificate to a Gnuk (version 1.2.6) device the command failed.

Same command works fine with "real" OpenPGP Cards.

### Steps to reproduce
Follow the steps from the [wiki](https://github.com/OpenSC/OpenSC/wiki/OpenPGP-card)
* grap yourself a cert or make one with `openssl req -x509 -newkey rsa:2048 -keyout key.pem -out cert.pem`
* try to import the cert on a Gnuk device (1.2.6) with `pkcs15-init --store-certificate cert.pem --id 3`

### Logs
```
$ pkcs15-init --store-certificate mycert.pem --id 3
Using reader with a card: Nitrokey Nitrokey Start (FSIJ-1.2.6-67111047) 00 00
Security officer PIN [Admin PIN] required.
Please enter Security officer PIN [Admin PIN]: 

(process:1949): GLib-GIO-CRITICAL **: g_application_send_notification: assertion '!g_application_get_is_remote (application)' failed
Failed to store certificate: Transmit failed
```
More detailed (-vvv):
https://gist.github.com/alex-nitrokey/05c00ea39988b93bf14fa40a7bbd1786

**Comment by Jakuje** [comment]
What OpenPGP versions are used in both of the applets? Do you test with opensc 0.17.0 or master?

**Comment by alex-nitrokey** [comment]
Sorry, right, important information...
GnuPG: 2.2.5
OpenSC: 0.17

Actually I did work with a git-version (not actual master), and downgraded to "normal" 0.17 after realizing this. It is the same behaviour though. System is Arch Linux. A user with similar recent programs (system unknown) [reported](https://support.nitrokey.com/t/unable-to-store-signed-certificate-on-nitrokey-start/971) this behaviour.

What do you mean with "in both of the applets"?

**Comment by erossi** [comment]
Hi,
I reported this, my working system is Debian Testing with opensc (0.17.0-3) GnuPG (2.2.5-1).

**Comment by Jakuje** [comment]
Maybe saying OpenPGP specification version in these applets would be more appropriate. You refer to Gnuk device (Nitrokey?) and "real" OpenPGP Card. Both of them have to implement some version of the specification. Until recent time, the versions 2 and 3 were not supported in OpenSC.

**Comment by alex-nitrokey** [comment]
Ah, I see. Both of them support OpenPGP Card version 2, did not test this behaviour on OpenPGP Card version 3 yet.

The Gnuk device (yes, Nitrokey Start) implements some functions of version 3 afaik, but officially is version 2.0 compatible (that's what gpg --card-status is stating).

**Comment by frankmorgner** [comment]
a failed transmit in general and your debug trace in particular indicates a hardware problem. It doesn't look like OpenSC could do a lot in this case. Is there enough space available to store the certificate?

**Comment by alex-nitrokey** [comment]
Maybe you are right, but I am not sure. If so, the wiki is inaccurate for a long time regarding the compatibility of Gnuk, but it is quite possible that no one realized yet.

What I experienced is, that using the command to import private key and certificate seems to work. (at least it is not different from what OpenPGP Card is doing) Therefore I would have thought, that Gnuk actually is capable to store the cert.

Please see the command [for key-cert pair import](https://github.com/OpenSC/OpenSC/wiki/OpenPGP-card#pairs-of-key--certificate-from-p12-file) and the corresponding output for [OpenPGP Card v2](https://gist.github.com/alex-nitrokey/5ec83459b37a9ea3981e91dca4c98ca7) (Nitrokey Pro) and [Gnuk 1.2.6](https://gist.github.com/alex-nitrokey/5469238b80c9bb07dbf32b0fc67f9f9b) (Nitrokey Start).

What do you think?

**Comment by frankmorgner** [comment]
Sorry, I don't know what's going on there.

If you see problems with the documentation, please update the description.

**Comment by alex-nitrokey** [comment]
Hey, long time no see...

After digging around in the source code I realized two points which may help to solve this issue:
* there is a special function for cert import on Gnuk devices [here](https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/card-openpgp.c#L1532)
* newer Gnuk (>1.0.x) devices are identified as "normal" OpenPGP_CARD_V2 devices and not as Gnuk device

There may be a connection. I try to further investigate the next weeks. Any hints are appreciated.

**Comment by dengert** [comment]
It sounds like you are seeing is a card vender writes their own card applet based on some standard but they  add some new feature, don't implement required features and/or don't follow the standard correctly. (For example Yubico with their PIV and OpenPGP applets.) 

Most OpenSC developers only have 1 or 2 different types of cards they can test with. So there is a very small number of developers (maybe only 1, you) that can reproduce your problem.  If the card you have has any additional command to identify itself, code could be added to handle the compatibility issues for that card. `grep CI_ card-piv.c`  for examples if issues. It also tries to read a Yubico version number, as different versions have different issues.

**Comment by jans23** [comment]
To import cerificates into Gnuk, Gnuk provides a dedicated Python script. The reason is that Gnuk doesn't follow the spec in this regards. 

I think the best solution would be to improve Gnuk so that it is compliant (instead of fixing OpenSC). However I don't know the technical constraints which yield to this inconsistency in the first place.

**Comment by alex-nitrokey** [comment]
Fixed by 3750d70 as key is detected as Gnuk again. Thus, the [special cert import function](https://github.com/OpenSC/OpenSC/blob/9d44adbc4e174f7cb562181a4bc5b003fbc039fb/src/libopensc/card-openpgp.c#L1589) is used again.
